notebooks/learning-to-rank/01-learning-to-rank.ipynb

```diff
@@ -0,0 +1,1556 @@
+              "2                        NaN     143.211  \n",
+              "3                        NaN      32.913  \n",
+              "4                        NaN      21.058  \n",
+              "...                      ...         ...  \n",
+              "384750                   NaN      68.287  \n",
+              "384751                   NaN      21.026  \n",
+              "384752                   NaN      32.913  \n",
+              "384753                   NaN      34.773  \n",
+              "384754                   NaN      25.920  \n",
+              "\n",
+              "[384755 rows x 8 columns]"
+            ]
+          },
+          "execution_count": 10,
+          "metadata": {},
+          "output_type": "execute_result"
+        }
+      ],
+      "source": [
+        "import numpy\n",
+        "\n",
+        "from eland.ml.ltr import FeatureLogger\n",
+        "\n",
+        "# First we create a feature logger that will be used to query Elasticsearch to retrieve the features:\n",
+        "feature_logger = FeatureLogger(es_client, MOVIE_INDEX, ltr_config)\n",
+        "\n",
+        "# This method will be applied for each group of query in the judgment log:\n",
+        "def _extract_query_features(query_judgements_group):\n",
+        "    # Retrieve document ids in the query group as strings.\n",
+        "    doc_ids = query_judgements_group['doc_id'].astype('str').to_list()\n",
+        "\n",
+        "    # Resolve query paras for the current query group (e.g.: {\"query\": \"batman\"}).\n",
+        "    query_params = { 'query': query_judgements_group['query'].iloc[0] }\n",
+        "\n",
+        "    # Extract the features for the documents in the query group:\n",
+        "    doc_features =  feature_logger.extract_features(query_params, doc_ids)\n",
+        "\n",
+        "    # Adding a column to the dataframe for each features:\n",
+        "    for feature_index, feature_name in enumerate(feature_logger._model_config.feature_names):\n",
+        "        query_judgements_group[feature_name] = numpy.array([doc_features[doc_id][feature_index] for doc_id in doc_ids])\n",
+        "\n",
+        "    return query_judgements_group\n",
+        "\n",
+        "judgments_with_features = judgments_df.groupby('query_id', group_keys=False).progress_apply(_extract_query_features)\n",
+        "\n",
+        "judgments_with_features"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 11,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "# This step will separate the dataset in two different parts one used for the training and one used for the evaluation of the model.\n",
+        "#\n",
+        "# We are not using sklearn.model_selection.train_test_split because it is ignoring query group during the split.\n",
+        "# In theory it should be possible to use it if you have enough pairs for each query in your judgment list.\n",
```

Comment:
Pull this out into a Markdown cell, it'll be easier to read I think

Reply:
This block does not exists anymore since I am now using another method to split the dataset as suggested by @maxjakob